src/stdlib/sieve.rs

```diff
@@ -0,0 +1,196 @@
+use std::collections::HashSet;
+
+use crate::compiler::prelude::*;
+
+fn sieve(
+    value: Value,
+    permitted_characters: Value,
+    replace_single: Value,
+    replace_repeated: Value,
+) -> Resolved {
+    let value = value.try_bytes_utf8_lossy()?;
+    let replace_single = replace_single.try_bytes_utf8_lossy()?;
+    let replace_repeated = replace_repeated.try_bytes_utf8_lossy()?;
+
+    match permitted_characters {
+        Value::Bytes(bytes) => {
+            let string = String::from_utf8_lossy(&bytes);
+            let characters: HashSet<char> = string.chars().collect();
+            let mut result = String::with_capacity(value.len());
+            let mut missed_length = 0;
+            for char in value.chars() {
+                if characters.contains(&char) {
+                    match missed_length {
+                        l if l > 1 => result += &replace_repeated,
+                        1 => result += &replace_single,
+                        _ => (),
+                    }
+                    missed_length = 0;
+                    result.push(char);
+                } else {
+                    missed_length += 1;
+                }
+            }
+            Ok(result.into())
+        }
+        Value::Regex(regex) => {
+            let mut result = String::with_capacity(value.len());
+            let mut last_end = 0;
+            for m in regex.find_iter(&value) {
+                match m.start() - last_end {
+                    l if l > 1 => result += &replace_repeated,
+                    1 => result += &replace_single,
+                    _ => (),
+                }
+                last_end = m.end();
+                result += m.as_str();
+            }
+            Ok(result.into())
+        }
+        value => Err(ValueError::Expected {
+            got: value.kind(),
+            expected: Kind::regex() | Kind::bytes(),
+        }
+        .into()),
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Sieve;
+
+impl Function for Sieve {
+    fn identifier(&self) -> &'static str {
+        "sieve"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "permitted_characters",
+                kind: kind::BYTES | kind::REGEX,
+                required: true,
+            },
+            Parameter {
+                keyword: "replace_single",
+                kind: kind::BYTES,
+                required: false,
+            },
+            Parameter {
+                keyword: "replace_repeated",
```

Comment:
❔ Can you explain a bit what this `replace_repeated` argument does? I think we could come up with a better name.

Reply:
Yeah, it is a bit confusing.

Based on the original issue (https://github.com/vectordotdev/vrl/issues/704), it is meant to provide a way to set a replacement string when multiple repeated characters are removed from the original string.

You can also see how it behaves in tests and examples.

`all_options` example might represent this the best:
```rust
all_options {
    args: func_args![value: value!("test123%456.فوائد.net."), permitted_characters: regex::Regex::new("[a-z.0-9]").unwrap(), replace_single: "X", replace_repeated: "<REMOVED>"],
    want: Ok(value!("test123X456.<REMOVED>.net.")),
    tdef: TypeDef::bytes().infallible(),
}
```

In this example the single `%` found in the string that was not allowed, was replaced with `X`, but that Arabic (maybe?) script string was replaced with `<REMOVED>`.